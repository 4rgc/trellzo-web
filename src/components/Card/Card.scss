@use 'card-variables' as v;

.card-base {
	display: flex;
	flex-direction: column;
	background-color: rgba($color: white, $alpha: 0.9);
	border-radius: 3px;
	margin: 0;
	overflow: hidden;
}

.card-cover {
	width: 100%;
	object-fit: cover;
	margin-bottom: 10px;
}

%card-cover-md,
%card-cover-sm {
	height: 50%;
}

%card-cover-lg {
	height: 30%;
}

%card-hover {
	cursor: pointer;
	background-color: rgba($color: #ddf, $alpha: 0.9);
}

%card-noimage {
	justify-content: center;
}

.card {
	@extend .card-base;

	@each $size, $value in v.$sizes {
		&.card-#{$size} {
			height: $value;
			.card-content-text {
				display: -webkit-box;
				-webkit-line-clamp: map-get($map: v.$line-clamp, $key: $size);
				-webkit-box-orient: vertical;
				overflow: hidden;
			}
			.card-cover {
				@extend %card-cover-#{$size};
			}
		}
		&.card-clickable {
			transition: background-color 0.3s ease;

			&:hover {
				@extend %card-hover;
			}
		}
		&.card-noimage {
			@extend %card-noimage;
		}
	}
}
.card-divider {
	margin: 0 auto;
	height: 0;
	width: 90%;
	border: 2px solid rgba(64, 182, 115, 0.7);
	border-radius: 5px;
}

.card-container {
	margin: 10px;
	h3 {
		margin: 4px 0;
		white-space: nowrap;
		overflow: hidden;
		text-overflow: ellipsis;
	}
	font-size: medium;

	&-no-text {
		@extend .card-container;
		overflow: hidden;
	}
}
