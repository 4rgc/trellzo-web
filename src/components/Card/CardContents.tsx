export interface ICardContentsProps {
	content?: string | React.ReactNode;
	title?: string;
	children?: React.ReactNode;
}

const CardContents: React.FC<ICardContentsProps> = ({
	content,
	title,
	children,
}) => (
	<div
		className={`card-container${
			typeof content !== 'string' ? '-no-text' : ''
		}`}
	>
		{children ?? (
			<>
				<h3 className="card-title">{title}</h3>
				<div
					className={
						typeof content === 'string' ? 'card-content-text' : ''
					}
				>
					{content}
				</div>
			</>
		)}
	</div>
);

export default CardContents;
