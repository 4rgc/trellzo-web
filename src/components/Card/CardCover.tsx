import { ReactComponent as Placeholder } from '../../assets/placeholder.svg';

export interface ICardCoverProps {
	src?: string;
}

const CardCover: React.FC<ICardCoverProps> = ({ src }) =>
	src ? (
		<img className="card-cover" src={src} alt="card img" />
	) : (
		<Placeholder
			fill="black"
			stroke="black"
			className="card-cover"
			preserveAspectRatio="xMinYMin slice"
		/>
	);

export default CardCover;
