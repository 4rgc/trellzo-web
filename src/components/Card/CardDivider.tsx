const CardDivider = () => <div className="card-divider"></div>;

export default CardDivider;
