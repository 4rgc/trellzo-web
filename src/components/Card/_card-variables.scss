$sizes: (
	'sm': 100px,
	'md': 300px,
	'lg': 500px,
);

$line-clamp: (
	'sm': 2,
	'md': 4,
	'lg': 14,
);
