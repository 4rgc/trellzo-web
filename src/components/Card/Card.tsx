import './Card.scss';
import { ReactComponent as Placeholder } from '../../assets/placeholder.svg';

interface ICardProps extends React.ComponentPropsWithoutRef<'div'> {
	onClick?: () => void;
	children?: React.ReactChild | React.ReactChild[];
	isDisabled?: boolean;
	imageUrl?: string;
	title?: string;
	content?: string | React.ReactElement | React.ReactElement[];
	size?: 'sm' | 'md' | 'lg';
}

const Card: React.FC<ICardProps> = (props) => {
	const {
		onClick,
		children,
		isDisabled = false,
		imageUrl,
		title,
		content,
		size = 'md',
		...otherProps
	} = props;

	return (
		<div className={`card-${size}`} onClick={onClick} {...otherProps}>
			{imageUrl ? (
				<img className="card-cover" src={imageUrl} alt="card img" />
			) : (
				<Placeholder
					fill="black"
					stroke="black"
					className="card-cover"
					preserveAspectRatio="xMinYMin slice"
				/>
			)}
			<div className="card-divider"></div>
			<div
				className={`card-container${
					typeof content !== 'string' ? '-no-text' : ''
				}`}
			>
				{children ?? (
					<>
						<h3>{title}</h3>
						<div
							className={
								typeof content === 'string'
									? 'card-content-text'
									: ''
							}
						>
							{content}
						</div>
					</>
				)}
			</div>
		</div>
	);
};

export default Card;
