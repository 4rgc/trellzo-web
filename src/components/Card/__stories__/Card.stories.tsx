import { ComponentStory, ComponentMeta } from '@storybook/react';

import Card from '../Card';

export default {
	title: 'Components/Card',
	component: Card,
	args: {
		content: `Labore odio voluptas voluptas vel qui vel. Aut est quo. Eveniet non tempore incidunt eum vero. Quia tempore error neque nam et numquam quidem. Amet voluptas delectus quisquam architecto numquam quia et impedit eos. Excepturi aliquid eum.`,
		title: 'Title',
	},
	argTypes: {
		onClick: { action: 'clicked' },
	},
} as ComponentMeta<typeof Card>;

const Template: ComponentStory<typeof Card> = ({ children, ...args }) => (
	<Card {...args} style={{ width: '200px' }}>
		{children}
	</Card>
);

export const Default = Template.bind({});

export const WithCover = Template.bind({});
WithCover.args = { imageUrl: '/regular-cover.png' };

export const WithVerticalCover = Template.bind({});
WithVerticalCover.args = { imageUrl: '/vertical-cover.png' };

export const WithHorizontalCover = Template.bind({});
WithHorizontalCover.args = { imageUrl: '/horizontal-cover.png' };

export const WithLongTitle = Template.bind({});
WithLongTitle.args = { title: 'Hello this is long title' };

export const WithCustomChildren: ComponentStory<typeof Card> = (args) => (
	<Card {...args} style={{ width: '200px' }}>
		<div
			style={{
				backgroundColor: 'blue',
				color: 'white',
				borderRadius: '10px',
				textAlign: 'center',
			}}
		>
			Custom stuff
		</div>
	</Card>
);

export const WithCustomContent: ComponentStory<typeof Card> = (args) => (
	<Card
		{...args}
		style={{ width: '200px' }}
		content={
			<div
				style={{
					backgroundColor: 'blue',
					color: 'white',
					borderRadius: '10px',
					textAlign: 'center',
				}}
			>
				Custom content
			</div>
		}
	/>
);
