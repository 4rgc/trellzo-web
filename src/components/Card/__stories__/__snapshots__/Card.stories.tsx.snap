// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Card Default smoke-test 1`] = `
<div class="card card-md card-clickable"
     style="width: 200px;"
>
  <svg viewbox="0 0 500 500"
       xmlns="http://www.w3.org/2000/svg"
       fill="black"
       stroke="black"
       class="card-cover"
       preserveaspectratio="xMinYMin slice"
  >
    <g transform="matrix(1, 0, 0, 1, 146.358994, 113.460999)"
       paint-order="fill markers stroke"
    >
      <path d="m118.507 133.514-34.249 34.249-15.968-15.968-41.938 41.937H178.726z"
            opacity="0.675"
            fill="#fff"
      >
      </path>
      <circle cx="58.217"
              cy="108.555"
              r="11.773"
              opacity="0.675"
              fill="#fff"
      >
      </circle>
      <path fill="none"
            d="M26.111 77.634h152.614v116.099H26.111z"
      >
      </path>
    </g>
  </svg>
  <div class="card-divider">
  </div>
  <div class="card-container">
    <h3 class="card-title">
      Title
    </h3>
    <div class="card-content-text">
      Labore odio voluptas voluptas vel qui vel. Aut est quo. Eveniet non tempore incidunt eum vero. Quia tempore error neque nam et numquam quidem. Amet voluptas delectus quisquam architecto numquam quia et impedit eos. Excepturi aliquid eum.
    </div>
  </div>
</div>
`;

exports[`Components/Card WithCover smoke-test 1`] = `
<div class="card card-md card-clickable"
     style="width: 200px;"
>
  <img class="card-cover"
       src="/regular-cover.png"
       alt="card img"
  >
  <div class="card-divider">
  </div>
  <div class="card-container">
    <h3 class="card-title">
      Title
    </h3>
    <div class="card-content-text">
      Labore odio voluptas voluptas vel qui vel. Aut est quo. Eveniet non tempore incidunt eum vero. Quia tempore error neque nam et numquam quidem. Amet voluptas delectus quisquam architecto numquam quia et impedit eos. Excepturi aliquid eum.
    </div>
  </div>
</div>
`;

exports[`Components/Card WithCustomChildren smoke-test 1`] = `
<div class="card card-md card-clickable"
     style="width: 200px;"
>
  <svg viewbox="0 0 500 500"
       xmlns="http://www.w3.org/2000/svg"
       fill="black"
       stroke="black"
       class="card-cover"
       preserveaspectratio="xMinYMin slice"
  >
    <g transform="matrix(1, 0, 0, 1, 146.358994, 113.460999)"
       paint-order="fill markers stroke"
    >
      <path d="m118.507 133.514-34.249 34.249-15.968-15.968-41.938 41.937H178.726z"
            opacity="0.675"
            fill="#fff"
      >
      </path>
      <circle cx="58.217"
              cy="108.555"
              r="11.773"
              opacity="0.675"
              fill="#fff"
      >
      </circle>
      <path fill="none"
            d="M26.111 77.634h152.614v116.099H26.111z"
      >
      </path>
    </g>
  </svg>
  <div class="card-divider">
  </div>
  <div class="card-container">
    <div style="background-color: blue; color: white; border-radius: 10px; text-align: center;">
      Custom stuff
    </div>
  </div>
</div>
`;

exports[`Components/Card WithCustomContent smoke-test 1`] = `
<div class="card card-md card-clickable"
     style="width: 200px;"
>
  <svg viewbox="0 0 500 500"
       xmlns="http://www.w3.org/2000/svg"
       fill="black"
       stroke="black"
       class="card-cover"
       preserveaspectratio="xMinYMin slice"
  >
    <g transform="matrix(1, 0, 0, 1, 146.358994, 113.460999)"
       paint-order="fill markers stroke"
    >
      <path d="m118.507 133.514-34.249 34.249-15.968-15.968-41.938 41.937H178.726z"
            opacity="0.675"
            fill="#fff"
      >
      </path>
      <circle cx="58.217"
              cy="108.555"
              r="11.773"
              opacity="0.675"
              fill="#fff"
      >
      </circle>
      <path fill="none"
            d="M26.111 77.634h152.614v116.099H26.111z"
      >
      </path>
    </g>
  </svg>
  <div class="card-divider">
  </div>
  <div class="card-container-no-text">
    <h3 class="card-title">
      Title
    </h3>
    <div style="background-color: blue; color: white; border-radius: 10px; text-align: center;">
      Custom content
    </div>
  </div>
</div>
`;

exports[`Components/Card WithHorizontalCover smoke-test 1`] = `
<div class="card card-md card-clickable"
     style="width: 200px;"
>
  <img class="card-cover"
       src="/horizontal-cover.png"
       alt="card img"
  >
  <div class="card-divider">
  </div>
  <div class="card-container">
    <h3 class="card-title">
      Title
    </h3>
    <div class="card-content-text">
      Labore odio voluptas voluptas vel qui vel. Aut est quo. Eveniet non tempore incidunt eum vero. Quia tempore error neque nam et numquam quidem. Amet voluptas delectus quisquam architecto numquam quia et impedit eos. Excepturi aliquid eum.
    </div>
  </div>
</div>
`;

exports[`Components/Card WithLongTitle smoke-test 1`] = `
<div class="card card-md card-clickable"
     style="width: 200px;"
>
  <svg viewbox="0 0 500 500"
       xmlns="http://www.w3.org/2000/svg"
       fill="black"
       stroke="black"
       class="card-cover"
       preserveaspectratio="xMinYMin slice"
  >
    <g transform="matrix(1, 0, 0, 1, 146.358994, 113.460999)"
       paint-order="fill markers stroke"
    >
      <path d="m118.507 133.514-34.249 34.249-15.968-15.968-41.938 41.937H178.726z"
            opacity="0.675"
            fill="#fff"
      >
      </path>
      <circle cx="58.217"
              cy="108.555"
              r="11.773"
              opacity="0.675"
              fill="#fff"
      >
      </circle>
      <path fill="none"
            d="M26.111 77.634h152.614v116.099H26.111z"
      >
      </path>
    </g>
  </svg>
  <div class="card-divider">
  </div>
  <div class="card-container">
    <h3 class="card-title">
      Hello this is long title
    </h3>
    <div class="card-content-text">
      Labore odio voluptas voluptas vel qui vel. Aut est quo. Eveniet non tempore incidunt eum vero. Quia tempore error neque nam et numquam quidem. Amet voluptas delectus quisquam architecto numquam quia et impedit eos. Excepturi aliquid eum.
    </div>
  </div>
</div>
`;

exports[`Components/Card WithVerticalCover smoke-test 1`] = `
<div class="card card-md card-clickable"
     style="width: 200px;"
>
  <img class="card-cover"
       src="/vertical-cover.png"
       alt="card img"
  >
  <div class="card-divider">
  </div>
  <div class="card-container">
    <h3 class="card-title">
      Title
    </h3>
    <div class="card-content-text">
      Labore odio voluptas voluptas vel qui vel. Aut est quo. Eveniet non tempore incidunt eum vero. Quia tempore error neque nam et numquam quidem. Amet voluptas delectus quisquam architecto numquam quia et impedit eos. Excepturi aliquid eum.
    </div>
  </div>
</div>
`;
